### PR TITLE
Adds example of overriding `setPassword` method

### PR DESCRIPTION
**Pull request by AP-Hunt**
Laravel Doctrine contributors,

I was implementing authentication using this library yesterday and, upon reading the docs, I wasn't clear where or when password hashing occurred when creating a user. In the end, it seemed to me the most sensible place was in the user model's setPassword method.

I decided to send a pull request with this little addition to the docs in the hopes of clearing this up for other users. If I've gone about password hashing all wrong, a point in the right direction would be gratefully received and I'll close this pull request -- I wouldn't want to point others in the same wrong direction I've gone!

Regards,
Andy

### CODE REVIEW

**Comment by josenicomaia** [comment]
Yeah, when changing the password, you have to set a hashed one.
But this way you are increasing the coupling in your model. I shouldn't recommend it.
There are some ways to solve it.

1 - Setting the hashed password instead of plain-text password
2 - Injecting an callback that will resolve the hashed password
3 - Injecting the service that can hash the password (use an interface for this)

## Example setting a hashed password:

> GenericController.php
```php
public class GenericController extends Controller {
  public function someMethod(Request $request) {
    $user = ... get user ...;
    $password = ... get password ...;
    $user->setPassword($hasher, Hash::make($password));
  }
}
```

> User.php
```php
use Illuminate\Contracts\Hashing\Hasher;

public class User {
  public function setPassword($hashedPassword) {
    $this->password = $hashedPassword;
  }
}
```

## Example using callback injection:

> GenericController.php
```php
public class GenericController extends Controller {
  public function someMethod(Request $request) {
    $user = ... get user ...;
    $password = ... get password ...;
    $user->setPassword(function () use ($password) {
      return Hash::make($password);
    });
  }
}
```

> User.php
```php
public class User {
  public function setPassword(callable $resolvePassword) {
    $this->password = $resolvePassword();
  }
}
```

## Example using service injection:

> GenericController.php
```php
use Illuminate\Contracts\Hashing\Hasher;

public class GenericController extends Controller {
  public function someMethod(Request $request, Hasher $hasher) {
    $user = ... get user ...;
    $password = ... get password ...;
    $user->setPassword($hasher, $password);
  }
}
```

> User.php
```php
use Illuminate\Contracts\Hashing\Hasher;

public class User {
  public function setPassword(Hasher $hasher, $plainTextPassword) {
    $this->password = $hasher->make($plainTextPassword);
  }
}
```

**Comment by AP-Hunt** [comment]
Hi José,

Thanks for getting back to me so quickly. I think I most agree with your third proposal (service injection) for the purposes of decoupling. However I think for sake of conciseness and clarity of documentation, having a direct call to the `Hash` facade is OK.

I'm not sure whether the aim is to guide other developers in how to structure their application, or simply to alert them to the fact they need to handle password hashing for themselves? (ie neither LaravelDoctrine not Laravel's authentication tools will handle this for you). If you feel the former is more important here, I'll amend my pull request to put more focus on the structural aspect.

Regards,
Andy

**Comment by patrickbrouwers** [comment]
Hey Andy,

Thanks for your PR. I think it's a good idea to add this information to the docs, however I don't feel the approach you proposed is the best way to go. 

I personally tackle password changes / registrations in a Service, as hashing isn't a Domain responsibility, but an application concern.  Whenever you need to change a Password (this can be through controller, console command, api, etc.) you would call that service. The service would encapsulate all the logic (making sure the password is correctly hashed, maybe some other business rules concerning password security). It will also remove the hard coupling between the Hasher/Laravel and your Domain.

Bonus point: it's easily testable!


Some pseudo code:
```
class PasswordService 
{

public __construct(Hasher $haser);

public function changePassword(User $user, $password)
{
        // start transaction

       $hashedPassword = $this->hasher->hash($password);

        $user->changePassword($hashedPassword);

         $this->repo->flush();

        // commit transaction
}

}
```

**Comment by AP-Hunt** [comment]
Hi Patrick,

I see the points you and José are making, and I agree that decoupling the two is important. I'll rework my PR tomorrow to try communicate this better, without being too verbose.

**Comment by patrickbrouwers** [comment]
Awesome, thanks man! Cheers :)

**Comment by AP-Hunt** [comment]
I've reworked the PR to address your concerns, José & Patrick. I felt that mentioning the initial, flawed implementation would be of help both to those looking for the absolute minimum they need to do*, as well as providing a nice contrast for the more decoupled approach.

(* I'm sure we've all been there: just needing to get the thing in with the bare minimum, to get a prototype working.)

File: orm/auth.md [hunk 1]
```diff
@@ -113,6 +113,56 @@ return [
...
+and LaravelDoctrine will treat passwords as nothing more than strings.
+
+The absolute minimum you can do to implement hashing is to override the
+`setPassword` method of the `\LaravelDoctrine\ORM\Auth\Authenticatable` trait.
+
+```
+use Illuminate\Support\Facades\Hash;
+class User implements \Illuminate\Contracts\Auth\Authenticatable
+{
+  use \LaravelDoctrine\ORM\Auth\Authenticatable;
```
Comment (josenicomaia) [inline orm/auth.md]: I guess that it should be 2 lines before

**Comment by josenicomaia** [comment]
I guess it's fine.
What do you think @patrickbrouwers ?

**Review by josenicomaia — CHANGES_REQUESTED** [review]
Ops. looks like that a forget to send the review.

**Comment by patrickbrouwers** [comment]
>  felt that mentioning the initial, flawed implementation would be of help both to those looking for the absolute minimum they need to do*, as well as providing a nice contrast for the more decoupled approach.

Not a big fan of that, I'd prefer not to have that kind of bad practice in the documentation.

File: orm/auth.md [hunk 2]
```diff
@@ -113,6 +113,56 @@ return [
...
+
+  public function setPassword($password)
+  {
+    $this->password = Hash::make($password);
```
Comment (patrickbrouwers) [inline orm/auth.md]: Remove this section

File: orm/auth.md [hunk 3]
```diff
@@ -113,6 +113,56 @@ return [
...
+However, we advise against this approach because it couples your domain entities
+to a specific hashing implementation (ie that provided by the `Hash` facade), as
+well as adding, implicitly, the responsibility of password hashing to your entity.
+Instead, we would advise a more decoupled and testable approach, such as
+implementing password-related operations as a service
+
+```
+use \Illuminate\Contracts\Hashing\Hasher;
+
+class PasswordService
```
Comment (patrickbrouwers) [inline orm/auth.md:126]: Can you make this code copy-and-pastable. Right now there're some problems.

**Comment by AP-Hunt** [comment]
@patrickbrouwers Appreciate your desire to keep it out of the docs. I've dropped that section and made the code more copy-paste friendly. I looked at other code samples in the repo to get a rough gauge of what language features were being used and what kind of detail to include in doc blocks, but let me know if you there's anything you want changing in it.

File: orm/auth.md [hunk 4]
```diff
@@ -113,6 +113,55 @@ return [
...
+   * Validate and change the given users password
+   *
+   * @param User $user
+   * @param string $password
+   * @throws PasswordTooWeakException
+   * @return void
+   */
+  public function changePassword(User $user, $password)
+  {
+    if($this->passwordStrengthValidator->isStrongEnough($password))
```
Comment (patrickbrouwers) [inline orm/auth.md]: Could you make this method (changePassword) PSR-2 (code style) compatible. Rest is looking good 👍

**Review by patrickbrouwers — CHANGES_REQUESTED** [review]

File: orm/auth.md (hunk 4 above)
Comment (AP-Hunt) [reply orm/auth.md]: Sorted. Made that whole block compliant. :+1:

**Comment by patrickbrouwers** [comment]
Looking good like that, thanks :)

**Pull request merged**